rholang/src/main/scala/coop/rchain/rholang/interpreter/accounting/CostAccounting.scala

```diff
@@ -7,54 +7,24 @@ import cats.mtl._
 import cats.{FlatMap, Monad}
 import coop.rchain.rholang.interpreter.errors.OutOfPhlogistonsError
 
-trait CostAccounting[F[_]] {
-  def charge(cost: Cost): F[Unit]
-  def get(): F[Cost]
-  def set(cost: Cost): F[Unit]
-  def refund(refund: Cost): F[Unit]
-  private[accounting] def update(f: Cost => Cost): F[Unit]
-}
-
 object CostAccounting {
-  def of[F[_]: Sync](init: Cost): F[CostAccounting[F]] =
+
+  def of[F[_]: Sync](init: Cost): F[MonadState[F, Cost]] =
     Ref[F].of(init).map(ref => new CostAccountingImpl[F](ref))
-  def empty[F[_]: Sync]: F[CostAccounting[F]] =
-    this.of(Cost(0, "init"))
 
-  def apply[F[_]](implicit ev: CostAccounting[F]): CostAccounting[F] = ev
+  def empty[F[_]: Sync]: F[MonadState[F, Cost]] =
+    this.of(Cost(0, "init"))
 
-  def unsafe[F[_]](init: Cost)(implicit F: Sync[F]): CostAccounting[F] = {
+  def unsafe[F[_]](init: Cost)(implicit F: Sync[F]): MonadState[F, Cost] = {
     val ref = Ref.unsafe[F, Cost](init)
     new CostAccountingImpl[F](ref)
   }
 
-  private class CostAccountingImpl[F[_]](state: Ref[F, Cost])(implicit F: Sync[F])
-      extends CostAccounting[F] {
-    override def charge(cost: Cost): F[Unit] = chargeInternal(_ - cost)
-
-    private def chargeInternal(f: Cost => Cost): F[Unit] =
-      for {
-        _ <- failOnOutOfPhlo
-        _ <- state.update(f)
-        _ <- failOnOutOfPhlo
-      } yield ()
-
-    override def get: F[Cost] = state.get
-    override def set(cost: Cost): F[Unit] =
-      state.set(cost)
-    override def update(f: Cost => Cost): F[Unit] = state.update(f)
-
-    override def refund(refund: Cost): F[Unit] = state.update(_ + refund)
-    private val failOnOutOfPhlo: F[Unit] =
-      FlatMap[F].ifM(state.get.map(_.value < 0))(F.raiseError(OutOfPhlogistonsError), F.unit)
-  }
-
-  implicit def costAccountingMonadState[F[_]: Monad](
-      costAccounting: CostAccounting[F]
-  ): MonadState[F, Cost] = new DefaultMonadState[F, Cost] {
+  private class CostAccountingImpl[F[_]: Monad](state: Ref[F, Cost])
```

Comment:
So, if we do `CostAccountingImpl[F[_]: Monad](state: F[Ref[F, Cost]]): MonadState[F, Cost] = ...` here, then we can simplify the constructors to just return `MonadState[F, Cost]`. That would allow us to use `loggingCost` without having to use `unsafeRunSync` to instantiate the cost account 👍 . See `MVarMonadState` in the shared folder.

Reply:
This code is outdated